pkg/cluster/updater.go

```diff
@@ -0,0 +1,372 @@
+	Events record.EventRecorder
+	// SelfObj is the runtime object for ourself to use for events.
+	SelfObj runtime.Object
+	// Config is the cluster config, which holds the version.
+	Config *v1.ConfigMap
+}
+
+// ComponentUpdater is responsible for updating
+// a single component in the cluster.
+// It takes the name of the component and a function
+// that will be used to update that component.
+// The name should be the name of the component
+// as it appears in the manifest file.
+type ComponentUpdater struct {
+	// Name is the name of the component to update.
+	Name string
+	// UpdateFn is the function used to update this component.
+	UpdateFn ComponentUpdaterFunc
+}
+
+// ContainerImage describes a container image. It holds
+// the repo / tag for the image.
+type ContainerImage struct {
+	// Repo is the repository this image is in.
+	Repo string
+	// Tag is the image tag.
+	Tag string
+}
+
+// String returns a stringified version of the Containerimage
+// in the format of $repo:$tag.
+func (ci *ContainerImage) String() string {
+	return fmt.Sprintf("%s:%s", ci.Repo, ci.Tag)
+}
+
+// UpdaterFunc is a function that can be used to update a
+// single component of a certain type. A concrete updater
+// function may be something like 'DaemonSetUpdaterFunc`.
+type ComponentUpdaterFunc func(clientset.Interface, string, *ContainerImage) error
+
+// NewClusterUpdater returns a ClusterUpdater struct with defaults.
+func NewClusterUpdater(client clientset.Interface, newVersion, oldVersion *ContainerImage, config *v1.ConfigMap) (*ClusterUpdater, error) {
+	b := record.NewBroadcaster()
+	r := b.NewRecorder(api.EventSource{
+		Component: "cluster-updater",
+	})
+	p, err := client.Core().Pods(api.NamespaceSystem).Get(os.Getenv("MY_POD_NAME"))
+	if err != nil {
+		return nil, fmt.Errorf("error getting self pod: %v", err)
+	}
+	return &ClusterUpdater{
+		Client:     client,
+		Components: DefaultComponentUpdaterList,
+		NewVersion: newVersion,
+		OldVersion: oldVersion,
+		RollbackFn: DefaultRollback,
+		Events:     r,
+		SelfObj:    p,
+		Config:     config,
+	}, nil
+}
+
+// DaemonsetRollingUpdate will perform a safe rolling update on the DaemonSet
+// specified by `name`. An error will be returned if the rolling update failed.
+func DaemonsetRollingUpdate(client clientset.Interface, name string, image *ContainerImage) error {
+	ds, err := client.Extensions().DaemonSets(api.NamespaceSystem).Get(name)
+	if err != nil {
+		return err
+	}
+	// Create new DS.
+	ds.Labels["version"] = image.Tag
+	ds.Spec.Template.Labels["version"] = image.Tag
+	for i, c := range ds.Spec.Template.Spec.Containers {
+		if c.Name == name {
+			glog.Infof("updating image for container: %s", c.Name)
+			ds.Spec.Template.Spec.Containers[i].Image = image.String()
+			break
+		}
+	}
+	ds, err = client.Extensions().DaemonSets(api.NamespaceSystem).Update(ds)
+	if err != nil {
+		return err
+	}
+	l, err := labels.Parse(fmt.Sprintf("k8s-app=%s", name))
```

Comment:
This is a bit fragile. We should be able to use the daemonset selector to get pods managed by this daemonset:

```
	selector, err := unversioned.LabelSelectorAsSelector(ds.Spec.Selector)
	if err != nil {
		return nil, err
	}
	daemonPods, err := client.Core().Pods(ds.Namespace).List(selector)
	if err != nil {
		return nil, err
	}
```